store/src/server/api.rs

```diff
@@ -188,6 +196,52 @@ impl api_server::Api for StoreApi {
             nullifiers: vec![],
         }))
     }
+
+    async fn get_transaction_inputs(
+        &self,
+        request: tonic::Request<GetTransactionInputsRequest>,
+    ) -> Result<Response<GetTransactionInputsResponse>, Status> {
+        let request = request.into_inner();
+
+        // scope to read from the accounts
+        let mut account_states = Vec::with_capacity(request.account_ids.len());
+        {
+            let accounts = self.account_tree.read().await;
+            for account_id in request.account_ids {
+                let account_hash = accounts.get_leaf(account_id.id).map_err(internal_error)?;
+                account_states.push(AccountTransactionInputRecord {
+                    account_id: Some(account_id),
+                    account_hash: Some(account_hash.into()),
+                });
+            }
+        }
+
+        // FIXME: race condition
+        // 1. account hashes is loaded above
+        // 2. a concurrent apply_block request is executed, updating the state
+        // 3. the load below happens, using state updated by 2, which doesn't match the block from 1
```

Comment:
I'm not sure if this has been discussed yet, but I believe there is a more inconspicuous race here. From what I understand, the transaction and block kernels are executed concurrently. So it is possible for the following to happen:

1. TXs 1..4 are put in a batch, using block `n`
2. The above batch is put into a block, `n+1`
3. TXs 5..8 are put in a batch, using block `n` instead of `n+1`
4. The changes from block `n+1` are applied to the store

In the example above, it is possible for a transaction from each batch to use the same nullifier.

Here are some possible solutions:

- Allow the above to happen, the block kernel would not apply the updates of the transactions that have been invalidated. The tradeoff is no need for synchronization, at the cost of not every proven transaction being useful (I assume this race would be rare)
- Change the semantics of the endpoint. Instead of returning "block in which a transaction consumed the nullifier", it would return "block in which the nullifier was marked to be consumed" (whatever block was used for the `get_transaction_inputs` before step 1 above).
- Delegate the responsibility of fixing this race to coordination in the txpool. (I think this is a poor strategy, because it won't work well in the future with a federated network)

Reply:
update: during our teamsync we discussed the above. and it should be a no issue. The txpool is not stateless. The stateview in the txpool keeps track of the transactions of an account (so, it knows the in-flight txs that change the account hash), and it keeps track of the nullifiers (so, it knows the nullifiers that are in the process of being consumed). and the pool has the responsibility of preventing the above from happening.